src/stylesheets/core/_functions.scss

```diff
@@ -1776,185 +2050,139 @@ system z-index
   $bg-grade: get-color-grade($bg-color);
   $our-color-tokens: ($preferred-link-color, $fallback-link-color, "ink");
 
+  $color-token-values: ();
+
   $link-token: false;
   $hover-token: false;
 
   @each $color-token in $our-color-tokens {
     //@debug "color token: " + $color-token;
-    // If the color token is a custom color, set a $custom flag
-    $custom: if(
-      type-of(map-get($assignments-theme-color, $color-token)) == "color",
-      true,
-      false
-    );
+    $custom: is-custom-color($color-token);
 
-    // Only get a link color if one has not yet been found
-    @if not $found {
-      $link-grade-token: get-color-grade($color-token);
-      $link-grade: if($link-grade-token < 10, 0, $link-grade-token);
-      $link-magic-number: abs($bg-grade - $link-grade);
-      $token-darker: false;
-      $token-lighter: false;
-      $link-family: false;
-      $link-vivid: false;
-      $hover-grade: false;
-      $hover-vivid: false;
-
-      // If the link color is custom, output theme tokens, not system tokens
-      @if $custom {
-        //@debug "uses custom color.";
-        $custom-token: $color-token;
-        $custom-token-lighter: false;
-        $custom-token-darker: false;
-        $custom-split: str-split($custom-token, "-");
-        $custom-grade: false;
-        $custom-grade-lighter: false;
-        $custom-grade-darker: false;
-        //@debug "custom split:" + $custom-split;
-
-        // set family as the first string in the split
-        $custom-family: nth($custom-split, 1);
-
-        // ignore vivid in token calculations, treat as default
-        @if index($custom-split, "vivid") {
-          $custom-split: remove($custom-split, "vivid");
-        }
+    $link-grade-token: get-color-grade($color-token);
+    $link-grade: $link-grade-token;
+    $link-magic-number: abs($bg-grade - $link-grade);
 
-        // set family and grade for "accent" families, since their family includes the split character
-        @if $custom-family == "accent" {
-          $custom-family: $custom-family + "-" + nth($custom-split, 2);
-          $custom-grade: if(
-            length($custom-split) == 3,
-            nth($custom-split, 3),
-            "default"
-          );
-        } @else {
-          $custom-grade: if(
-            length($custom-split) == 2,
-            nth($custom-split, 2),
-            "default"
-          );
-        }
```

Comment:
Separated the custom color code into their own functions w/ documentation so it's easier to understand and debug in the future.

- get-custom-color-shades()
  - is-custom-color()
  - get-custom-color-split()
  - get-custom-color-family()
  - get-custom-color-grade()
  - get-custom-color-darker-grade()
  - get-custom-color-lighter-grade()